volatility3/framework/plugins/linux/lsof.py

```diff
@@ -1,12 +1,12 @@
 # This file is Copyright 2024 Volatility Foundation and licensed under the Volatility Software License 1.0
 # which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
 #
-"""A module containing a collection of plugins that produce data typically
-found in Linux's /proc file system."""
-import logging, datetime
-from typing import List, Callable
+import logging
+from datetime import datetime
+from dataclasses import dataclass, astuple, field
```

Comment:
I know these are built-in, but please can we stick to the style guide we use, which says all imports should be modules rather than methods, classes or objects please (`typing` gets a free pass)?  This prevents people importing `lsof.astuple` and thinking that's where it's defined.  See https://google.github.io/styleguide/pyguide.html#22-import

This also applies to datetime which was previously correctly imported.

Reply:
hm sorry but that's the Google style guide.. is it mentioned anywhere that we stick to Google's? why?

We should stick to the Python's styles guide which is the PEP8 ... [PEP 8 – Style Guide for Python Code](https://peps.python.org/pep-0008/#imports)  , for which that's perfectly fine.. actually there is an example that shows that's correct:
```
# Correct:
from subprocess import Popen, PIPE
```

Also for classes:
```
When importing a class from a class-containing module, it’s usually okay to spell this:

from myclass import MyClass
from foo.bar.yourclass import YourClass

If this spelling causes local name clashes, then spell them explicitly:

import myclass
import foo.bar.yourclass

and use myclass.MyClass and foo.bar.yourclass.YourClass.
```

Reply:
It's actually form the [Volatility style guide](https://github.com/volatilityfoundation/volatility/wiki/Style-Guide) from 2014 which is mostly in alignment with PEP8 with some extra requests.  I thought that was a bit terse and difficult to read, so I pointed you at Google's style guide which is longer with more discussion, but here's the original text:

* Importing specific objects or functions is discouraged.
** It pollutes the namespace and causes confusion. Previous versions of volatility allowed people to import functions from files they weren't defined in.
* "from blah import *" is strongly discouraged.
** Again, namespace pollution and inappropriate imports.

The entire rest of the code base is formatted that way and it's better to have a style guide than not.  I honestly didn't expect this request to be a big sticking point?  Can you explain why allowing namespace pollution is a better stylistic option than our current one, or is your disagreement just that the documentation was provided by Google and that code comes out slightly longer?

Reply:
No worries. I will take care of this.

I'm not saying it's better, but it's unclear to me why this is a concern now. I've never heard it mentioned in any of my previous contributions or by others. 

We'll also need to fix this, if we want to make an exception with Typing, in about 50 other imports. Otherwise, there are 244 imports to address.

Also, Volatility2 is a separate project, so we can't expect people to refer to it for guidance on writing code for Volatility3. Volatility3 should have its own dedicated style guide page, or clear instructions of which style guide we use.

I wonder if there's a way to include that as a rule in Black. Any idea? By the way, Black follows Python's PEP8 style guide.

I also learned we have a [.style.yapf ](https://github.com/volatilityfoundation/volatility3/blob/develop/.style.yapf) file, which configures the [Yapf Python formatted from Google](https://github.com/google/yapf). 

The **.style.yapf** file doesn't specify any `based_on_style`, which means it's defaulting to **PEP8** as the style guide. However, it seems we can switch it to Google's style guide if preferred. See [formatting-style](https://github.com/google/yapf?tab=readme-ov-file#formatting-style)

```yaml
based_on_style = google
```

Maybe we should use that instead of **Black** and try to check/fix this type of  issues too.
Otherwise, it might be a good idea to remove that file, as it creates confusion about which formatter the project actually uses.

Reply:
You can go back through previous pull requests.  Where I've spotted it I should have always asked for it to be corrected (and provided the rationale).  Most people stick with it, I'm disappointed I missed it for 244 imports!  5:(  It's possible I had quibbled a lot about the PRs and was feeling sorry for them having to fix so many things, but I feel it's saved us for lots of cross-imports between plugins and the core so I think it's a worthwhile guideline to have.

Yep, fair point, the style guide never got pulled across to vol 3 I don't think and I didn't want to sit down and rewrite it again (I didn't even tidy up the one we had), but we should.

I don't think black has that as a flag (I don't think black has many flags, to be fair).  If you find one, it would be great to apply it to our automatic black check, so that it's applied to all PRs evenly when people submit them!

Yeah, that's the style mechanism we used very early on, it was deprecated years ago but I guess I didn't remove the file.  I've taken care of that now.  Yapf was very bad, it would reweight something, and then because of the reweighting running it again would change it back, meaning it never reached a stable equilibrium.  It's the reason we moved to black.

We don't use the Google style guide exclusively, but there are a couple of points in it that make sense, such as only importing modules (in fact, that's the main one I can think of), which we've carried over.  As mentioned the main style guide is in general PEP 8.